faer-libs/faer-core/src/complex_native/c32_impl.rs

```diff
@@ -0,0 +1,1051 @@
+        self.re *= rhs;
+        self.im *= rhs;
+    }
+}
+
+impl core::ops::MulAssign for c32 {
+    #[inline(always)]
+    fn mul_assign(&mut self, rhs: c32) {
+        let tmp = self.re;
+        self.re *= rhs.re;
+        self.re -= rhs.im * self.im;
+        self.im *= rhs.re;
+        self.im += rhs.im * tmp;
+    }
+}
+
+impl core::ops::MulAssign<c32conj> for c32 {
+    #[inline(always)]
+    fn mul_assign(&mut self, rhs: c32conj) {
+        let tmp = self.re;
+        self.re *= rhs.re;
+        self.re += rhs.neg_im * self.im;
+        self.im *= rhs.re;
+        self.im -= rhs.neg_im * tmp;
+    }
+}
+
+impl core::ops::DivAssign<f32> for c32 {
+    #[inline(always)]
+    fn div_assign(&mut self, rhs: f32) {
+        self.re /= rhs;
+        self.im /= rhs;
+    }
+}
+
+impl core::ops::DivAssign for c32 {
+    #[inline(always)]
+    fn div_assign(&mut self, rhs: c32) {
+        *self *= rhs.faer_inv();
+    }
+}
+
+impl core::ops::DivAssign<c32conj> for c32 {
+    #[inline(always)]
+    fn div_assign(&mut self, rhs: c32conj) {
+        *self *= rhs.canonicalize().faer_inv();
+    }
+}
+
+impl core::ops::RemAssign<f32> for c32 {
+    #[inline(always)]
+    fn rem_assign(&mut self, rhs: f32) {
+        self.re %= rhs;
+        self.im %= rhs;
+    }
+}
+
+impl core::ops::RemAssign for c32 {
+    #[inline(always)]
+    fn rem_assign(&mut self, rhs: c32) {
+        *self = *self % rhs;
+    }
+}
+
+impl core::ops::RemAssign<c32conj> for c32 {
+    #[inline(always)]
+    fn rem_assign(&mut self, rhs: c32conj) {
+        *self = *self % rhs;
+    }
+}
+
+impl num_traits::Inv for c32 {
+    type Output = c32;
+
+    #[inline(always)]
+    fn inv(self) -> Self::Output {
+        self.faer_inv()
+    }
+}
+
+impl num_traits::Num for c32 {
+    type FromStrRadixErr =
+        num_complex::ParseComplexError<<f32 as num_traits::Num>::FromStrRadixErr>;
+
+    #[inline(always)]
+    fn from_str_radix(str: &str, radix: u32) -> Result<Self, Self::FromStrRadixErr> {
+        let num_complex = num_complex::Complex32::from_str_radix(str, radix)?;
+        Ok(num_complex.into())
+    }
+}
+
+impl From<c32> for num_complex::Complex32 {
+    #[inline(always)]
+    fn from(value: c32) -> Self {
+        Self {
+            re: value.re,
+            im: value.im,
+        }
+    }
+}
+
+impl From<num_complex::Complex32> for c32 {
+    #[inline(always)]
+    fn from(value: num_complex::Complex32) -> Self {
+        c32 {
+            re: value.re,
+            im: value.im,
+        }
+    }
+}
+
+impl From<f32> for c32 {
+    #[inline(always)]
+    fn from(value: f32) -> Self {
+        Self::new(value, 0.0)
+    }
+}
+
+impl<'a> From<&'a f32> for c32 {
+    #[inline(always)]
+    fn from(value: &'a f32) -> Self {
+        Self::new(value.clone(), 0.0)
+    }
+}
+
+unsafe impl bytemuck::Zeroable for c32 {}
+unsafe impl bytemuck::Pod for c32 {}
+
+impl core::fmt::Debug for c32 {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
```

Comment:
what we could do here is extract the sign bit and absolute value of `self.im`, and then branch based on the sign bit